ddtrace/propagation/http.py

```diff
@@ -238,48 +283,23 @@ def _extract(headers):
         )
 
         meta = None
-        tags_value = _extract_header_value(
-            _POSSIBLE_HTTP_HEADER_TAGS,
-            headers,
-            default="",
-        )
+
+        tags_value = _DatadogMultiHeader._get_tags_value(headers)
         if tags_value:
-            # Do not fail if the tags are malformed
-            try:
-                meta = {
-                    k: v
-                    for (k, v) in decode_tagset_string(tags_value).items()
-                    if (
-                        k not in _DatadogMultiHeader._X_DATADOG_TAGS_EXTRACT_REJECT
-                        and _DatadogMultiHeader._is_valid_datadog_trace_tag_key(k)
-                    )
-                }
-            except TagsetMaxSizeDecodeError:
-                meta = {
-                    "_dd.propagation_error": "extract_max_size",
-                }
-                log.warning("failed to decode x-datadog-tags", exc_info=True)
-            except TagsetDecodeError:
-                meta = {
-                    "_dd.propagation_error": "decoding_error",
-                }
-                log.debug("failed to decode x-datadog-tags: %r", tags_value, exc_info=True)
+            meta = _DatadogMultiHeader._extract_meta(tags_value)
 
+        # When 128 bit trace ids are propagated the 64 lowest order bits are set in the `x-datadog-trace-id`
+        # header. The 64 highest order bits are encoded in base 16 and store in the `_dd.p.tid` tag.
+        # Here we reconstruct the full 128 bit trace_id if 128-bit trace id generation is enabled.
         if meta and _HIGHER_ORDER_TRACE_ID_BITS in meta:
-            # When 128 bit trace ids are propagated the 64 lowest order bits are set in the `x-datadog-trace-id`
-            # header. The 64 highest order bits are encoded in base 16 and store in the `_dd.p.tid` tag.
-            # Here we reconstruct the full 128 bit trace_id.
             trace_id_hob_hex = meta[_HIGHER_ORDER_TRACE_ID_BITS]
-            try:
-                if len(trace_id_hob_hex) != 16:
-                    raise ValueError("Invalid size")
-                # combine highest and lowest order hex values to create a 128 bit trace_id
-                trace_id = int(trace_id_hob_hex + "{:016x}".format(trace_id), 16)
-            except ValueError:
+            if _DatadogMultiHeader._higher_order_is_valid(trace_id_hob_hex):
+                if config._128_bit_trace_id_enabled:
+                    trace_id = _DatadogMultiHeader._put_together_trace_id(trace_id_hob_hex, trace_id)
+            else:
                 meta["_dd.propagation_error"] = "malformed_tid {}".format(trace_id_hob_hex)
+                del meta[_HIGHER_ORDER_TRACE_ID_BITS]
                 log.warning("malformed_tid: %s. Failed to decode trace id from http headers", trace_id_hob_hex)
-            # After the full trace id is reconstructed this tag is no longer required
-            del meta[_HIGHER_ORDER_TRACE_ID_BITS]
```

Comment:
My thought process was that if it's correct, why would we not want it on the span? We need to have the tag on the span because it's tested in system-tests etc. Before my plan was to remove it as a tag entirely and always generate it from the full trace_id value (how I originally implemented it). However, that won't work with how tests are currently written. I also didn't see it as much of an issue having it set on the span. 

Let me know if that seems incorrect though, happy to discuss.